examples/vae/train_vae.py

```diff
@@ -119,39 +121,65 @@ def main():
          'main/reconstr', 'main/kl_penalty', 'elapsed_time']))
     trainer.extend(extensions.ProgressBar())
 
+    # If snapshot file is given, resume the training.
     if args.resume is not None:
         chainer.serializers.load_npz(args.resume, trainer)
 
     # Run the training
     trainer.run()
 
-    # Visualize the results
-    def save_images(x, filename):
-        import matplotlib.pyplot as plt
-        fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
-        for ai, xi in zip(ax.flatten(), x):
-            ai.imshow(xi.reshape(28, 28))
-        fig.savefig(filename)
+    # Save images for demonstration
+    save_images(device, encoder, decoder, train, test, prior, args.out)
 
-    avg_elbo_loss.to_cpu()
-    train_ind = [1, 3, 5, 10, 2, 0, 13, 15, 17]
-    x = chainer.Variable(np.asarray(train[train_ind]))
-    with chainer.using_config('train', False), chainer.no_backprop_mode():
-        x1 = decoder(encoder(x).mean, inference=True).mean
-    save_images(x.array, os.path.join(args.out, 'train'))
-    save_images(x1.array, os.path.join(args.out, 'train_reconstructed'))
 
-    test_ind = [3, 2, 1, 18, 4, 8, 11, 17, 61]
-    x = chainer.Variable(np.asarray(test[test_ind]))
-    with chainer.using_config('train', False), chainer.no_backprop_mode():
-        x1 = decoder(encoder(x).mean, inference=True).mean
-    save_images(x.array, os.path.join(args.out, 'test'))
-    save_images(x1.array, os.path.join(args.out, 'test_reconstructed'))
+# Saves 3x3 tiled image
+def save3x3(x, filename):
+    numpy_device = chainer.get_device('@numpy')
+    fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
+    for ai, xi in zip(ax.flatten(), x):
+        im = xi.reshape(28, 28)
+        im = numpy_device.send(im)
+        ai.imshow(im)
+    fig.savefig(filename)
+
 
-    # draw images from randomly sampled z
+# Saves reconstruction images using:
+# - training image samples
+# - test image samples
+# - randomly sampled values of z
+def save_images(device, encoder, decoder, train, test, prior, out_dir):
+
+    # Training samples
+    train_ind = [1, 3, 5, 10, 2, 0, 13, 15, 17]
+    x_numpy = train[train_ind]
+    x = np.asarray(device.send(x_numpy))
```

Comment:
`device.send(numpy.asarray(train[train_ind]))`?

Reply:
Thanks, that's better. Fixed.